LIFX light effects

amelchio opened the pull request:
**Update for users:** if you know how, please test this :)

**Update for reviewers:** the below description covers the first three commits. By quite popular demand I have now added a `colorloop` effect as well. That took some work, and the PR is now quite big. Let me know if you want some reading guidance.

## Description:

This is my first stab at LIFX light effects support. These are intended as short-running effects that are useful for notifications.

Maybe the implementation is controversial, I don’t know :-)

I have chosen not to build on the `flash` or `effect` parameters of `light.turn_on` and `light.turn_off` because I think those calls are already too overloaded (I can elaborate if you like). Adding in different LIFX specific operations would just make matters even worse.

So I wanted a new service call, but not a generic one like `light.effect_start`.

The term "light effects" sounds like something collective. In reality, though, each light effect has its own distinct set of parameters. For example, potential `blink`, `fireplace` and `party` effects could have quite different options. Thus, I think it is best to make each effect have its own service call. This avoids the situation of having a bunch of parameters where an unpredictable subset will not make sense in any given situation. It also makes it easier for users to discover all effects and their parameters. 

Consequently, even though the two initial effects do share the same parameter set, I have split them into two service calls: `light.lifx_effect_breathe` and `light.lifx_effect_pulse`. This matches the API chosen by LIFX, so there is some familiarity for LIFX users: https://api.developer.lifx.com/docs/breathe-effect and https://api.developer.lifx.com/docs/pulse-effect

I have added these service calls only for `lifx` because I think it is infeasible to implement them without hardware support.

Documentation will follow once the implementation settles down.

## Example entry for `configuration.yaml`:
```yaml
automation:
  - alias: Forgot to close window
    trigger:
      platform: state
      entity_id: binary_sensor.aeotec_zw112_door_window_sensor_6_sensor_1_0
      from: 'off'
      to: 'on'
      for:
        minutes: 30
    action:
      - service: light.lifx_effect_pulse
        data:
          entity_id: light.office, light.kitchen
          color_name: 'blue'
          brightness: 200
          period: 0.25
          cycles: 3
```

## Checklist:

If user exposed functionality or configuration variables are added/changed:
  - [ ] Documentation added/updated in [home-assistant.github.io](https://github.com/home-assistant/home-assistant.github.io)

If the code communicates with devices, web services, or third-party tools:
  - [X] Local tests with `tox` run successfully.

amelchio committed:
Add AwaitAioLIFX

This encapsulates the callback and Event that aiolifx needs and thus avoids an
explosion of those when new calls are added.

The refresh_state is now generally useful, so move it into its own method.

amelchio committed:
Refactor into find_hsbk

This will be useful for new methods that also have to find passed in colors.

amelchio committed:
Initial effects support for LIFX

These effects are useful as notifications. They mimic the breathe and pulse
effects from the LIFX HTTP API:

    https://api.developer.lifx.com/docs/breathe-effect
    https://api.developer.lifx.com/docs/pulse-effect

However, this implementation runs locally with the LIFX LAN protocol.

cmsimike commented:
This looks like it is using the lan protocol (rather than going through the lifx server) for the effects?

amelchio commented:
@cmsimike That is correct.

cmsimike commented:
@amelchio thank you!!

amelchio committed:
Organize LIFX effects in classes

This is to move the setup/restore away from the actual effect, making it quite
simple to add additional effects.

amelchio committed:
Stop running LIFX effects on conflicting service calls

Turning the light on/off or starting a new effect will now stop the running
effect.

amelchio committed:
Present default LIFX effects as light.turn_on effects

This makes the effects (with default parameters) easily accessible from
the UI.

amelchio committed:
Add LIFX colorloop effect

This cycles the HSV colors, so that is added as an internal way to set a
color.

amelchio committed:
Saturate LIFX no color value

Now the color is "full saturation, no brightness". This avoids a lot of
temporary white when fading from the "no color" value and into a real color.

balloob reviewed (approved):
This looks great 🐬 

Before we merge, may I suggest one change: create a lifx folder, rename `lift.py` to `lifx/__init__.py` and break out the effects into a new file. Might make it all a bit more readable.

amelchio commented:
Thanks. I will make that change tonight (CET), hoping to get the PR included in the next release. Would make it easier to get some testers 😺 

Do I put it all in `lifx/__init__.py` or should I create a separate file for each effect?

balloob commented:
I would just create 1 file for the effects.

amelchio committed:
Move lifx to its own package and split effects into a separate file

amelchio committed:
Always show LIFX light name in logs

The name is actually the easiest way to identify a bulb so just using it
as a fallback was a bit odd.

amelchio committed:
Compact effect getter

amelchio committed:
Always use full brightness for random flash color

This is a stopgap. When a bit more infrastructure is in place, the intention
is to turn the current hue some degrees. This will guarantee a flash color
that is both unlike the current color and unlike white.

amelchio committed:
Clear effects concurrently

We have to wait for the bulbs, so let us wait for all of them at once.

amelchio committed:
Add lifx_effect_stop

The colorloop effect is most impressive if run on many lights. Testing
this has revealed the need for an easy way to stop effects on all lights
and return to the initial state of each bulb. This new call does just that.

Calling turn_on/turn_off could also stop the effect but that would not
restore the initial state.

amelchio committed:
Always calculate the initial effect color

To fade nicely from power off, the breathe effect needs to keep an
unchanging hue. So give up on using a static start color and just find the
correct hue from the target color.

The colorloop effect can start from anything but we use a random color
just to keep things a little interesting during power on.

amelchio committed:
Fix lint

amelchio committed:
Update .coveragerc

amelchio commented:
@balloob I have moved the effects into their own file now. It’s much better that way, thanks! I also did a little further tweaking so I think it is good to go now.

balloob commented:
Cherry-picked for 0.43
